### PR TITLE
[DESK-296][DESK-375] Graphql API integration

### PR DESCRIPTION
**Pull request by JamieRuderman**
### Description

Change the device list query from the old REST API to use the new GraphQL API.

### Checklist

- [x] Pull Request title is in the format `[PROJ-12345] Some useful description here...`
- [x] The git branch is in the format `PROJ-12345-some-useful-description-here`
- [x] I have added one or more reviewers
- [x] This PR only contains one isolated change (bug fix, feature, etc)
- [x] I have manually reviewed this PR to make sure only the files I intended to change were changed and nothing else
- [x] I have notified reviewers on Slack by @ mentioning them in the `#desktop`channel
- [x] I have moved the Jira ticket into `Resolved` state and made a note in the ticket with a link to this PR

### Test plan

1. Sign in
2. Go to the device list
3. Search, sort, filter, load more. 
4. You should see it work for all size device lists even with 1000 devices

### Ticket

<https://remoteit.atlassian.net/browse/DESK-296>
<https://remoteit.atlassian.net/browse/DESK-375>

### CODE REVIEW

File: frontend/src/models/devices.ts [hunk 1]
```diff
@@ -1,213 +1,122 @@
...
-      // Handle unset state - localStorage turns bool into string
-      if (typeof pref === 'string') {
-        searchOnly = pref === 'true'
+    async fetch(_, globalState: any) {
+      const { set, graphQLProcessor } = dispatch.devices
+      const { query, filter, size, from, append } = globalState.devices
+      const options: gqlOptions = {
+        size,
+        from,
+        state: filter === 'all' ? '' : filter,
```
Comment (penngrove) [inline frontend/src/models/devices.ts:44]: I would do:

state: filter === 'all' ? undefined : filter,

File: frontend/src/models/devices.ts [hunk 2]
```diff
@@ -1,213 +1,122 @@
...
+        name: query.length > 1 ? query : '',
```
Comment (penngrove) [inline frontend/src/models/devices.ts:45]: same here with undefined

File: frontend/src/models/devices.ts [hunk 3]
```diff
@@ -1,213 +1,122 @@
...
-
-      // Handle unset state - localStorage turns bool into string
-      if (typeof pref === 'string') {
-        searchOnly = pref === 'true'
+    async fetch(_, globalState: any) {
+      const { set, graphQLProcessor } = dispatch.devices
+      const { query, filter, size, from, append } = globalState.devices
+      const options: gqlOptions = {
+        size,
+        from,
```
Comment (penngrove) [inline frontend/src/models/devices.ts:43]: Maybe use the infinite pagination with last/after

{
  login {
    devices(size: 5, after: "80:00:00:00:01:04:8B:59") {
      hasMore
      last
      items {
        name
      }
    }
  }
}

File: frontend/src/pages/ServicesPage/ServicesPage.tsx [hunk 4]
```diff
@@ -45,30 +46,45 @@ export const ServicesPage = connect(mapState)(({ connections, devices }: Service
...
       }
     >
-      {/* <Typography variant="subtitle1">Services</Typography> */}
+      {searched && <Subtitle primary="Services" secondary={`Searched for “${query}”`} />}
       <ServiceList services={device.services} connections={serviceConnections} />
       <Typography variant="subtitle1">Device details</Typography>
       <Columns count={1} inset>
         <DataDisplay
           data={[
+            { label: 'Availability', value: percent(device.availability) },
```
Comment (penngrove) [inline frontend/src/pages/ServicesPage/ServicesPage.tsx:61]: This is a percent average online/duration

File: frontend/src/pages/ServicesPage/ServicesPage.tsx [hunk 5]
```diff
@@ -45,30 +46,45 @@ export const ServicesPage = connect(mapState)(({ connections, devices }: Service
...
+            { label: 'Instability', value: percent(device.instability) },
```
Comment (penngrove) [inline frontend/src/pages/ServicesPage/ServicesPage.tsx:62]: Average number of state cycles per day

Comment (penngrove) [reply frontend/src/pages/ServicesPage/ServicesPage.tsx:62]: So not a percent

File: frontend/src/pages/ServicesPage/ServicesPage.tsx [hunk 6]
```diff
@@ -45,30 +46,45 @@ export const ServicesPage = connect(mapState)(({ connections, devices }: Service
...
+  if (date) return <Duration startTime={date.getTime()} ago />
+}
+
+function percent(value: number) {
+  if (value) return Math.round(value) + '%'
+}
+
+function location(geo: IDevice['geo']) {
+  if (!geo) return null
+  let loc = isNaN(+geo.state) ? geo.state + ' ' : ''
```
Comment (penngrove) [inline frontend/src/pages/ServicesPage/ServicesPage.tsx:88]: City is more cool

Comment (penngrove) [reply frontend/src/pages/ServicesPage/ServicesPage.tsx:88]: State is the first subdivision in https://en.wikipedia.org/wiki/ISO_3166-2

File: frontend/src/pages/ServicesPage/ServicesPage.tsx [hunk 7]
```diff
@@ -45,30 +46,45 @@ export const ServicesPage = connect(mapState)(({ connections, devices }: Service
...
+      {searched && <Subtitle primary="Services" secondary={`Searched for “${query}”`} />}
       <ServiceList services={device.services} connections={serviceConnections} />
       <Typography variant="subtitle1">Device details</Typography>
       <Columns count={1} inset>
         <DataDisplay
           data={[
+            { label: 'Availability', value: percent(device.availability) },
+            { label: 'Instability', value: percent(device.instability) },
+            { label: 'Location', value: location(device.geo) },
+            { label: 'ISP', value: device.geo?.isp },
```
Comment (penngrove) [inline frontend/src/pages/ServicesPage/ServicesPage.tsx:64]: Add connection type: One of the following values: Dialup, Cable/DSL, Corporate, or Cellular. Additional values may be added in the future.

File: frontend/src/pages/ServicesPage/ServicesPage.tsx (hunk 6 above)
Comment (penngrove) [inline frontend/src/pages/ServicesPage/ServicesPage.tsx:88]: Will get you the name

File: frontend/src/services/graphQL.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import axios from 'axios'
+import { r3 } from '../services/remote.it'
+import { parseType } from '../services/serviceTypes'
+import { renameServices } from '../helpers/nameHelper'
+import { GRAPHQL_API_URL } from '../constants'
+import { updateConnections } from '../helpers/connectionHelper'
+
+export async function graphQLFetch({ size, from, state = '', name = '', ids = [] }: gqlOptions) {
+  const connections = `connections: devices(id:${JSON.stringify(ids)})`
+  const devices = `devices(size:${size}, from:${from}, name:"${name}", state: "${state}")`
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:10]: Use variables

Comment (penngrove) [reply frontend/src/services/graphQL.ts:10]: Maybe use after/last

File: frontend/src/services/graphQL.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  const select = `
+  {
+    total
+    items {
+      id
+      name
+      state
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:17]: would prefer endpoint.state

File: frontend/src/services/graphQL.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+          country
+          state
+          isp
+        }
+      }
+      owner {
+        id
+        email
+      }
+      services {
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:37]: Warning, bulk service may disappear

File: frontend/src/services/graphQL.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        name
+        id
+        port
+        type
+        created
+        endpoint {
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:43]: use state to be consistent

File: frontend/src/services/graphQL.ts [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+      query: `
+        {
+          login {
+            id
+            ${devices}${select}
+            ${connections}${select}
+          }
+        }
+      `,
+    },
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:64]: add variables here

Comment (penngrove) [reply frontend/src/services/graphQL.ts:64]: without $

File: frontend/src/services/graphQL.ts [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+  })
+}
+
+export function graphQLAdaptor(gqlDevices: any, loginId: string, hidden?: boolean): IDevice[] {
+  if (!gqlDevices) return []
+  let data = gqlDevices?.items.map(
+    (d: any): IDevice => {
+      const services = d.services.reduce((result: IService[], s: any): IService[] => {
+        const { typeID, type } = parseType(s.type)
+        if (typeID !== 35)
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:74]: will be done in the backend

File: frontend/src/services/graphQL.ts [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+      }
+      owner {
+        id
+        email
+      }
+      services {
+        name
+        id
+        port
+        type
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:41]: application (which is the number part like 35 for bulk)

                "type": "00:1C:00:00:00:04:00:06:04:30:00:16:00:01:00:00",
                "application": 28,
                "title": "SSH"

File: frontend/src/services/graphQL.ts [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    },
+  })
+}
+
+export function graphQLAdaptor(gqlDevices: any, loginId: string, hidden?: boolean): IDevice[] {
+  if (!gqlDevices) return []
+  let data = gqlDevices?.items.map(
+    (d: any): IDevice => {
+      const services = d.services.reduce((result: IService[], s: any): IService[] => {
+        const { typeID, type } = parseType(s.type)
```
Comment (penngrove) [inline frontend/src/services/graphQL.ts:73]: can get typeid from service.application

File: frontend/src/models/devices.ts (hunk 1 above)
Comment (JamieRuderman) [reply frontend/src/models/devices.ts:44]: changed

File: frontend/src/models/devices.ts (hunk 2 above)
Comment (JamieRuderman) [reply frontend/src/models/devices.ts:45]: removed the length check. will move to search form.

**Pull request merged**